2.3-TapTree.ipynb

```diff
@@ -0,0 +1,500 @@
+    "assert(taproot_script2 == taproot_script)\n",
+    "print(taptree.desc)\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 1C - Huffman Constructor\n",
+    "\n",
+    "A TapTree can also be constructed with a greedy algorithm based on the huffman encoding algorithm. \n",
+    "Simply provide the huffman constructor method with the relative probability weights of each TapLeaf.\n",
+    "\n",
+    "* **A higher weight** \n",
+    "    * Indicates a higher likelihood of execution.\n",
+    "    * Means the script will be placed closer to the root if possible.\n",
+    "    * Results in a smaller inclusion proof and lower spending fees.\n",
+    "\n",
+    "![test](images/TapTree3.jpg)\n",
+    "\n",
+    "\n",
+    "#### Example 1.3 - Building a TapTree with the huffman constructor.\n",
+    "* We reconstruct the same taptree from examples 1.1 and 1.2 with the huffman constructor."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Run example 1.1 and 1.2 first.\n",
+    "\n",
+    "taptree = TapTree()\n",
+    "taptree.key = pk_internal\n",
+    "taptree.huffman_constructor([(1, tapleaf_A), (1, tapleaf_B), (2, tapleaf_C)])\n",
+    "print(taptree.desc)\n",
+    "\n",
+    "# Compare with example 1.1\n",
+    "taproot_script3, tweak3, control_map3 = taptree.construct()\n",
+    "assert(taproot_script3 == taproot_script)\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Part 2: Spending along the Script Path\n",
+    "\n",
+    "A Taproot output is spent along the script path with the following witness pattern.\n",
+    "\n",
+    "* Witness to spend TapScript_A:\n",
+    "\n",
+    "    * `[Stack element(s) satisfying TapScript_A]`\n",
+    "    * `[TapScript_A]` \n",
+    "    * `[Controlblock c]`\n",
+    "\n",
+    "\n",
+    "* Controlblock c contains:\n",
+    "\n",
+    "    * `[Tapscript Version]` \n",
+    "        * `0xfe & c[0]`\n",
+    "    * `[Internal Public Key]` \n",
+    "        * y-coordinate: `0x01 & c[0]` \n",
+    "        * x-coordinate: `c[1:33]`\n",
+    "    * `[Script Inclusion Proof]` \n",
+    "        * `n x 32Bytes`\n",
+    "\n",
+    "\n",
+    "_Note: As of writing, the Taproot BIP has been updated to reflect 32B public keys in the controlblock. Following examples and exercises continue to use 33B public keys until the taproot Bitcoin Core branch has been updated._\n",
+    "\n",
+    "![test](images/TapTree4.jpg)\n",
+    "\n",
+    "\n",
+    "**Generating the Controlblock**\n",
+    "\n",
+    "We use the the TapTree construct method to generate the taproot output, tweak and controlblocks for all TapScripts.\n",
+    "\n",
+    "**`TapTree.construct()` returns the tuple:**\n",
+    "* `taproot_output_script`, `tweak`, `control_block_map`\n",
+    "* Control block map has key-value pairs: `tapscript` - `controlblock`\n",
```

Comment:
suggest s/Control block map/`control_block_map` for clarity

Reply:
done